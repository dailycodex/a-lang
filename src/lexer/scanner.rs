use super::Span;
use crate::parse::{
    keyword,
    CtrlColon,
    CtrlComma,
    CtrlDot,
    CtrlLBrace,
    CtrlLBracet,
    CtrlLParan,
    CtrlRBrace,
    CtrlRBracet,
    CtrlRParan,
    CtrlRightArrow,
    // CtrlStar,
    // CtrlSlash,
    CtrlSemiColon,
    Ident,
    LitBool,
    LitChar,
    LitInt,
    LitStr,
    OpAdd,
    OpDiv,
    OpEqual,
    OpEqualEqual,
    OpGeq,
    OpGrt,
    OpLeq,
    OpLes,
    OpMul,
    OpNeq,
    OpNot,
    OpSub,
};
use std::iter::Peekable;
use std::str::Chars;

type Token = Box<dyn super::Token>;

pub struct Lexer<'a> {
    src: Peekable<Chars<'a>>,
    span: Span,
    last_chr_len: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Self {
            src: src.chars().peekable(),
            span: Span::default(),
            last_chr_len: 0,
        }
    }

    fn peek(&mut self) -> Option<&char> {
        self.src.peek()
    }

    fn next(&mut self) -> Option<char> {
        let Some(ch) = self.src.next() else {
            return None;
        };
        self.span.right_shift(ch);
        self.last_chr_len = ch.to_string().as_bytes().len();
        Some(ch)
    }

    fn next_if<F>(&mut self, func: F) -> Option<char>
    where
        F: FnOnce(char) -> bool,
    {
        let Some(c) = self.peek() else {
            return None;
        };
        if func(*c) {
            return self.next();
        }
        None
    }

    fn span(&mut self) -> Span {
        let span = self.span;
        self.span.reset(None);
        span
    }

    fn number(&mut self, c: char) -> Option<Token> {
        let mut number = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_digit() || c == '_') {
            number.push(c);
        }
        Some(Box::new(LitInt::new(number, self.span())))
    }

    fn ident(&mut self, c: char) -> Option<Token> {
        let mut id = c.to_string();
        while let Some(c) = self.next_if(|c| c.is_ascii_alphanumeric() || c == '_') {
            id.push(c);
        }
        let span = self.span();
        Some(match id.as_str() {
            "fn" => Box::new(keyword::Fn(span)),
            "struct" => Box::new(keyword::Struct(span)),
            "if" => Box::new(keyword::If(span)),
            "else" => Box::new(keyword::Else(span)),
            "use" => Box::new(keyword::Use(span)),
            "return" => Box::new(keyword::Return(span)),
            "let" => Box::new(keyword::Let(span)),
            "true" => Box::new(LitBool::new(id, span)),
            "false" => Box::new(LitBool::new(id, span)),
            _ => Box::new(Ident::new(id, span)),
        })
    }

    fn string(&mut self) -> Option<Token> {
        let mut string = String::new();
        while let Some(c) = self.next_if(|c| c != '"') {
            string.push(c);
        }
        self.next();
        Some(Box::new(LitStr::new(string, self.span())))
    }

    fn chr(&mut self) -> Option<Token> {
        let mut string = String::new();
        while let Some(c) = self.next_if(|c| c != '\'') {
            string.push(c);
        }
        self.next();

        Some(Box::new(LitChar::new(string, self.span())))
    }
    fn take_while(&mut self, expected: char) {
        while self.next_if(|c| c != expected).is_some() {}
    }

    fn comment(&mut self) -> Option<Token> {
        self.take_while('\n');
        let Some(ch) = self.next() else {
            return None;
        };
        self.parse(ch)
    }

    fn token<T>(&mut self, op: &str) -> Option<Token>
    where
        T: super::Token,
    {
        for _ in 0..op.chars().count().saturating_sub(self.last_chr_len) {
            self.next();
        }
        Some(Box::new(T::new(op.into(), self.span())))
    }

    fn matched(&mut self, ch: char) -> bool {
        matches!(self.peek(), Some(c) if c == &ch)
    }

    fn parse(&mut self, ch: char) -> Option<Token> {
        match ch {
            n @ '0'..='9' => self.number(n),
            i @ ('a'..='z' | 'A'..='Z') => self.ident(i),
            '"' => self.string(),
            '\'' => self.chr(),
            '/' if self.matched('/') => self.comment(),
            '-' if self.matched('>') => self.token::<CtrlRightArrow>("->"),
            '>' if self.matched('=') => self.token::<OpGeq>(">="),
            '<' if self.matched('=') => self.token::<OpLeq>("<="),
            '=' if self.matched('=') => self.token::<OpEqualEqual>("=="),
            '!' if self.matched('=') => self.token::<OpNeq>("!="),
            '-' => self.token::<OpSub>("-"),
            '+' => self.token::<OpAdd>("+"),
            '*' => self.token::<OpMul>("*"),
            '/' => self.token::<OpDiv>("/"),
            '>' => self.token::<OpGrt>(">"),
            '<' => self.token::<OpLes>("<"),
            '=' => self.token::<OpEqual>("="),
            '!' => self.token::<OpNot>("!"),
            // '%' => self.op_token("%"),
            '.' => self.token::<CtrlDot>("."),
            ',' => self.token::<CtrlComma>(","),
            '(' => self.token::<CtrlLParan>("("),
            ')' => self.token::<CtrlRParan>(")"),
            '{' => self.token::<CtrlLBrace>("{"),
            '}' => self.token::<CtrlRBrace>("}"),
            '[' => self.token::<CtrlLBracet>("{"),
            ']' => self.token::<CtrlRBracet>("}"),
            ':' => self.token::<CtrlColon>(":"),
            ';' => self.token::<CtrlSemiColon>(";"),
            // 'λ' => self.op_token("λ"),
            '\n' | '\r' | ' ' | '\0' => {
                let Some(ch) = self.next() else {
                    return None;
                };
                self.span.reset(Some(self.last_chr_len));
                self.parse(ch)
            }
            _ => panic!("unknown char {ch:?}"),
        }
    }

    pub fn lex(mut self) -> Result<Vec<Token>, Vec<String>> {
        let mut tokens = vec![];
        while let Some(ch) = self.next() {
            let Some(token) = self.parse(ch) else {
                break;
            };
            tokens.push(token);
        }
        Ok(tokens)
    }
}
