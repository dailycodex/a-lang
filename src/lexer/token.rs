use super::Span;
use std::any::Any;
use std::fmt::Debug;

pub trait Token: Any + Debug {
    fn new(value: String, span: Span) -> Self
    where
        Self: Sized;
    fn value(&self) -> String;
    fn span(&self) -> Span;
    fn as_any(&self) -> &dyn Any;
}
