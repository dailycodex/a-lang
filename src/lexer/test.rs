use crate::lexer::lex;

pub fn snapshot_lexing(input: &str) -> String {
    let tokens = match lex(input) {
        Ok(l) => l,
        Err(e) => return e.into_iter().map(|i| format!("{i}\n")).collect(),
    };
    let mut tokens = std::collections::VecDeque::from(tokens.stream);
    let mut output = String::new();
    for (row, line) in input.lines().enumerate() {
        output += line;
        output += "\n";
        while let Some(tok) = tokens.pop_front() {
            // if tok.span().row_start != tok.span().row_end {
            //     panic!("We haven't handled this yet");
            // }

            if tok.span().row_end != row {
                tokens.push_front(tok);
                break;
            }

            output += &" ".repeat(tok.span().col_start);
            output += &"^".repeat(tok.span().len()); // tok.span().col_end - tok.span().col_start);
            output += &format!(" {tok:?}");
            output += "\n"
        }
    }

    output
}

macro_rules! snapshot {
    ($name:tt, $path:tt) => {
        #[test]
        fn $name() {
            let contents = include_str!($path);
            let mut settings = insta::Settings::clone_current();
            settings.set_snapshot_path("testdata/output/");
            settings.bind(|| {
                insta::assert_snapshot!(snapshot_lexing(contents));
            });
        }
    };
}

snapshot!(binary, "testdata/snapshots/binary.a");
snapshot!(ifelse, "testdata/snapshots/ifelse.a");
