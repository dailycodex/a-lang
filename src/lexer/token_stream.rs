use crate::lexer::Token;

#[derive(Debug)]
pub struct TokenStream {
    pub(crate) stream: Vec<Box<dyn Token>>,
    idx: usize,
}

impl TokenStream {
    pub(crate) fn new(stream: Vec<Box<dyn Token>>) -> Self {
        Self { stream, idx: 0 }
    }

    pub fn is_not_at_end(&self) -> bool {
        self.idx < self.stream.len() - 1
    }

    // pub fn next(&mut self) -> Option<&dyn Token> {
    //     let result = self.stream.get(self.idx).map(|t| &**t)
    //     self.idx += 1;
    //     result
    // }

    pub fn next_as<Expected>(&mut self) -> Option<&Expected>
    where
        Expected: Token,
    {
        let result = self
            .stream
            .get(self.idx)
            .and_then(|i| i.as_any().downcast_ref::<Expected>());
        self.idx += 1;
        result
    }

    pub fn next_if<Expected>(&mut self) -> Option<&Expected>
    where
        Expected: Token,
    {
        let result = self
            .stream
            .get(self.idx)
            .and_then(|i| i.as_any().downcast_ref::<Expected>());
        // if result.is_none() {
        //     return None;
        // }
        result?;
        self.idx += 1;
        result
    }

    // pub fn previous<'a>(&'a mut self) -> Option<&'a Box<dyn Token>> {
    //     if self.idx == 0 {
    //         return None;
    //     }
    //     self.stream.get(self.idx.saturating_sub(1))
    // }

    pub fn is_peek_a<Expected: 'static>(&self) -> bool
    where
        Expected: Token,
    {
        self.stream
            .get(self.idx)
            .and_then(|i| i.as_any().downcast_ref::<Expected>())
            .is_some()
    }

    pub fn peek_blind(&mut self) -> Option<&dyn Token> {
        self.stream.get(self.idx).map(|i| &**i)
    }

    pub fn peek<Expected>(&self) -> Option<&Expected>
    where
        Expected: Token,
    {
        self.stream
            .get(self.idx)
            .and_then(|i| i.as_any().downcast_ref::<Expected>())
    }

    // pub fn peek_nth<'a, Expected>(&'a self, nth: usize) -> Option<&'a Expected>
    // where
    //     Expected: Token,
    // {
    //     self.stream
    //         .get(self.idx + nth.saturating_sub(1))
    //         .and_then(|i| i.as_any().downcast_ref::<Expected>())
    // }

    // pub fn reset(&mut self) {
    //     self.idx = 0;
    // }
}

// #[test]
// fn test_peek() {
//     use super::lex;
//     use crate::lexer::Span;
//     use crate::parse::{LitInt, OpAdd};
//     use pretty_assertions::assert_eq;
//     let tokens = lex("1 + 1").unwrap();
//     assert!(tokens.is_peek_a::<LitInt>());
//     assert_eq!(
//         tokens.peek::<LitInt>(),
//         Some(&LitInt::new("1", Span::new(0, 0, 1)))
//     );
//     assert_eq!(
//         tokens.peek_nth::<OpAdd>(2),
//         Some(&OpAdd::new("+", Span::new(0, 2, 3)))
//     );
//     assert_eq!(
//         tokens.peek_nth::<LitInt>(3),
//         Some(&LitInt::new("1", Span::new(0, 4, 5)))
//     );
// }
