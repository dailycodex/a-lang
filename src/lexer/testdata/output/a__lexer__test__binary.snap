---
source: src/lexer/test.rs
expression: snapshot_lexing(contents)
---
fn main() {
^^ Fn((0,0)->(0,2))
   ^^^^ Ident 'main' (0,3)->(0,7)
       ^ CtrlLParan '(' (0,7)->(0,8)
        ^ CtrlRParan ')' (0,8)->(0,9)
          ^ CtrlLBrace '{' (0,10)->(0,11)
  1 + 3;
  ^ LitInt '1' (1,2)->(1,3)
    ^ OpAdd '+' (1,4)->(1,5)
      ^ LitInt '3' (1,6)->(1,7)
       ^ CtrlSemiColon ';' (1,7)->(1,8)
}
^ CtrlRBrace '}' (2,0)->(2,1)

